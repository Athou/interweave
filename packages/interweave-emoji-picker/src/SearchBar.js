/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import React from 'react';
import PropTypes from 'prop-types';
import { SEARCH_THROTTLE } from './constants';

type SearchBarProps = {
  autoFocus: boolean,
  onChange: (query: string) => void,
  query: string,
};

type SearchBarState = {
  query: string,
};

export default class SearchBar extends React.PureComponent<SearchBarProps, SearchBarState> {
  input: ?HTMLInputElement;
  timeout: ?number;

  static contextTypes = {
    classNames: PropTypes.objectOf(PropTypes.string),
    messages: PropTypes.objectOf(PropTypes.string),
  };

  static propTypes = {
    autoFocus: PropTypes.bool.isRequired,
    query: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
  };

  constructor({ query }: SearchBarProps) {
    super();

    this.state = {
      query,
    };
  }

  /**
   * Focus the input field if `autoFocus` is true.
   */
  componentDidMount() {
    if (this.props.autoFocus && this.input) {
      this.input.focus();
    }
  }

  /**
   * When the parent `Picker` search query is reset, also reset the input field.
   */
  componentWillReceiveProps({ query }: SearchBarProps) {
    if (query === '') {
      this.setState({
        query,
      });
    }
  }

  /**
   * Triggered when the input field value changes.
   */
  handleChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const query = e.currentTarget.value;

    // Update the input field immediately
    this.setState({
      query,
    });

    // But defer filtering in the picker
    clearTimeout(this.timeout);

    this.timeout = setTimeout(() => {
      this.props.onChange(query);
    }, SEARCH_THROTTLE);
  };

  /**
   * Set input field as reference.
   */
  handleRef = (ref: ?HTMLInputElement) => {
    this.input = ref;
  };

  render() {
    const { classNames, messages } = this.context;

    return (
      <div className={classNames.search}>
        <input
          aria-label={messages.searchAria}
          className={classNames.searchInput}
          placeholder={messages.search}
          ref={this.handleRef}
          type="search"
          value={this.state.query}
          onChange={this.handleChange}
        />
      </div>
    );
  }
}
