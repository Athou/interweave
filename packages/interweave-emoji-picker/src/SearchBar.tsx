/**
 * @copyright   2016, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import withContext, { ContextProps } from './Context';
import { SEARCH_THROTTLE } from './constants';
import { ContextShape } from './shapes';

export interface SearchBarProps {
  autoFocus: boolean;
  onChange: (query: string, event: React.ChangeEvent<HTMLInputElement>) => void;
  onKeyUp: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  searchQuery: string;
}

export interface SearchBarState {
  query: string;
}

export class SearchBar extends React.PureComponent<SearchBarProps & ContextProps, SearchBarState> {
  static propTypes = {
    autoFocus: PropTypes.bool.isRequired,
    context: ContextShape.isRequired,
    onChange: PropTypes.func.isRequired,
    onKeyUp: PropTypes.func.isRequired,
    searchQuery: PropTypes.string.isRequired,
  };

  inputRef = React.createRef<HTMLInputElement>();

  state = {
    query: this.props.searchQuery,
  };

  /**
   * Focus the input field if `autoFocus` is true.
   */
  componentDidMount() {
    if (this.props.autoFocus && this.inputRef.current) {
      this.inputRef.current.focus();
    }
  }

  /**
   * When the parent `Picker` search query is reset, also reset the input field.
   */
  componentDidUpdate(prevProps: SearchBarProps) {
    if (this.props.searchQuery === '' && prevProps.searchQuery) {
      this.setState({
        query: '',
      });
    }
  }

  /**
   * Triggered when the input field value changes.
   */
  private handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.persist();

    // Update the input field immediately
    this.setState({
      query: event.currentTarget.value,
    });

    // But defer filtering in the picker
    this.handleChangeDebounced(event);
  };

  /**
   * A change handler that is debounced for performance.
   */
  private handleChangeDebounced = debounce(event => {
    // console.log(this.inputRef.current, event.target.value, event.currentTarget.value);
    // Check if were still mounted
    if (this.inputRef.current) {
      this.props.onChange(event.currentTarget.value.trim(), event);
    }
  }, SEARCH_THROTTLE);

  render() {
    const { classNames, messages } = this.props.context;
    const { onKeyUp } = this.props;

    return (
      <div className={classNames.search}>
        <input
          aria-label={messages.searchA11y}
          className={classNames.searchInput}
          placeholder={messages.search}
          ref={this.inputRef}
          type="search"
          value={this.state.query}
          onChange={this.handleChange}
          onKeyUp={onKeyUp}
        />
      </div>
    );
  }
}

export default withContext(SearchBar);
