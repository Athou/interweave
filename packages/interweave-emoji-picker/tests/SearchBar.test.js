import React from 'react';
import { shallow, mount } from 'enzyme';
import SearchBar from '../src/SearchBar';
import { PICKER_CONTEXT } from './mocks';

jest.mock('lodash/debounce', () => jest.fn(fn => fn));

describe('<SearchBar />', () => {
  const context = PICKER_CONTEXT;

  const props = {
    autoFocus: false,
    onChange() {},
    onKeyUp() {},
    searchQuery: '',
  };

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders a search bar', () => {
    const wrapper = shallow(<SearchBar {...props} />, { context });

    expect(wrapper).toMatchSnapshot();
  });

  it('focuses on mount if `autoFocus` is true', () => {
    const wrapper = mount(<SearchBar {...props} />, {
      context,
      disableLifecycleMethods: true,
    });

    const spy = jest.spyOn(wrapper.instance().input, 'focus');

    wrapper.instance().componentDidMount();

    expect(spy).not.toHaveBeenCalled();

    wrapper.setProps({
      autoFocus: true,
    });

    wrapper.instance().componentDidMount();

    expect(spy).toHaveBeenCalled();
  });

  it('resets query if prop changes', () => {
    const wrapper = shallow(<SearchBar {...props} searchQuery="foo" />, { context });

    expect(wrapper.state('query')).toBe('foo');
    expect(wrapper.find('input').prop('value')).toBe('foo');

    wrapper.setProps({
      searchQuery: '',
    });

    expect(wrapper.state('query')).toBe('');
    expect(wrapper.find('input').prop('value')).toBe('');
  });

  it('can customize class name', () => {
    const wrapper = shallow(<SearchBar {...props} />, {
      context: {
        ...context,
        classNames: {
          ...context.classNames,
          search: 'test-search',
          searchInput: 'test-search-input',
        },
      },
    });

    expect(wrapper.prop('className')).toBe('test-search');
    expect(wrapper.find('input').prop('className')).toBe('test-search-input');
  });

  it('can customize messages', () => {
    const wrapper = shallow(<SearchBar {...props} />, {
      context: {
        ...context,
        messages: {
          ...context.messages,
          search: 'search',
          searchA11y: 'searchA11y',
        },
      },
    });

    expect(wrapper.find('input').prop('placeholder')).toBe('search');
    expect(wrapper.find('input').prop('aria-label')).toBe('searchA11y');
  });

  it('triggers `onKeyUp` when pressing keys', () => {
    const spy = jest.fn();
    const wrapper = shallow(<SearchBar {...props} onKeyUp={spy} />, { context });

    wrapper.find('input').simulate('keyup');

    expect(spy).toHaveBeenCalled();
  });

  it('triggers `onChange` when changing values', () => {
    const spy = jest.fn();
    const wrapper = mount(<SearchBar {...props} onChange={spy} />, { context });

    wrapper.find('input').simulate('change', {
      persist() {},
      target: {
        value: 'foo',
      },
    });

    expect(spy).toHaveBeenCalledWith('foo', expect.objectContaining({}));
  });

  it('sets changed value to state', () => {
    const wrapper = shallow(<SearchBar {...props} />, { context });

    expect(wrapper.state('query')).toBe('');

    wrapper.find('input').simulate('change', {
      persist() {},
      target: {
        value: 'bar',
      },
    });

    expect(wrapper.state('query')).toBe('bar');
  });

  it('trims changed value', () => {
    const spy = jest.fn();
    const wrapper = mount(<SearchBar {...props} onChange={spy} />, { context });

    wrapper.find('input').simulate('change', {
      persist() {},
      target: {
        value: ' baz ',
      },
    });

    expect(spy).toHaveBeenCalledWith('baz', expect.objectContaining({}));
  });
});
