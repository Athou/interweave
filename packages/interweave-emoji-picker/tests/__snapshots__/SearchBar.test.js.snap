// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> renders a search bar 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBar
    autoFocus={false}
    onChange={[Function]}
    onKeyUp={[Function]}
    searchQuery=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <input
        aria-label="Search for emojis by keyword"
        className="interweave-picker__search-input"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Search"
        type="text"
        value=""
      />,
      "className": "interweave-picker__search",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-label": "Search for emojis by keyword",
        "className": "interweave-picker__search-input",
        "onChange": [Function],
        "onKeyUp": [Function],
        "placeholder": "Search",
        "type": "text",
        "value": "",
      },
      "ref": [Function],
      "rendered": null,
      "type": "input",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <input
          aria-label="Search for emojis by keyword"
          className="interweave-picker__search-input"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />,
        "className": "interweave-picker__search",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-label": "Search for emojis by keyword",
          "className": "interweave-picker__search-input",
          "onChange": [Function],
          "onKeyUp": [Function],
          "placeholder": "Search",
          "type": "text",
          "value": "",
        },
        "ref": [Function],
        "rendered": null,
        "type": "input",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "classNames": Object {
        "emoji": "interweave-picker__emoji",
        "emojiActive": "interweave-picker__emoji--active",
        "emojis": "interweave-picker__emojis",
        "emojisBody": "interweave-picker__emojis-body",
        "emojisContainer": "interweave-picker__emojis-container",
        "emojisHeader": "interweave-picker__emojis-header",
        "emojisRow": "interweave-picker__emojis-row",
        "emojisSection": "interweave-picker__emojis-section",
        "group": "interweave-picker__group",
        "groupActive": "interweave-picker__group--active",
        "groups": "interweave-picker__groups",
        "groupsList": "interweave-picker__groups-list",
        "noPreview": "interweave-picker__no-preview",
        "noResults": "interweave-picker__no-results",
        "picker": "interweave-picker__picker",
        "pickerVirtual": "interweave-picker__picker--virtual",
        "preview": "interweave-picker__preview",
        "previewContent": "interweave-picker__preview-content",
        "previewEmoji": "interweave-picker__preview-emoji",
        "previewSubtitle": "interweave-picker__preview-subtitle",
        "previewTitle": "interweave-picker__preview-title",
        "search": "interweave-picker__search",
        "searchInput": "interweave-picker__search-input",
        "skinTone": "interweave-picker__skin-tone",
        "skinToneActive": "interweave-picker__skin-tone--active",
        "skinTones": "interweave-picker__skin-tones",
      },
      "messages": Object {
        "activities": "Activities",
        "animalsNature": "Animals & Nature",
        "flags": "Flags",
        "foodDrink": "Food & Drink",
        "frequentlyUsed": "Frequently Used",
        "noGroups": "All Emojis",
        "noPreview": "No emoji to display",
        "noResults": "No results",
        "objects": "Objects",
        "recentlyUsed": "Recently Used",
        "search": "Search",
        "searchA11y": "Search for emojis by keyword",
        "searchResults": "Search Results",
        "smileysPeople": "Smileys & People",
        "symbols": "Symbols",
        "travelPlaces": "Travel & Places",
      },
    },
  },
}
`;
